src/Servant/Multipart.hs

```diff
@@ -267,6 +301,100 @@ instance ( FromMultipart tag a
                     $ lookupContext popts config
       subserver' = addMultipartHandling pbak multipartOpts subserver
 
+-- | Upon seeing @MultipartForm a :> ...@ in an API type,
+--   servant-client will take a parameter of type @(LBS.ByteString, a)@,
+--   where the bytestring is the boundary to use (see 'genBoundary'), and
+--   replace the request body with the contents of the form.
+instance (ToMultipart tag a, HasClient m api, MultipartBackend tag)
+      => HasClient m (MultipartForm tag a :> api) where
+
+  type Client m (MultipartForm tag a :> api) = 
+    (LBS.ByteString, a) -> Client m api
+
+  clientWithRoute pm _ req (boundary, param) =
+      clientWithRoute pm (Proxy @api) $ setRequestBody newBody newMedia req
+    where
+      newBody = multipartToBody boundary $ toMultipart @tag param
+      newMedia = "multipart" // "form-data" /: ("boundary", LBS.toStrict boundary)
+
+  hoistClientMonad pm _ f cl = \a ->
+      hoistClientMonad pm (Proxy @api) f (cl a)
+
+-- | Generates a boundary to be used to separate parts of the multipart.
+-- Requires 'IO' because it is randomized.
+genBoundary :: IO LBS.ByteString
+genBoundary = LBS.pack 
+            . foldr (\x acc -> validChars ! x : acc) [] 
+            <$> indices
+  where
+    -- the standard allows up to 70 chars, but most implementations seem to be
+    -- in the range of 40-60, so we pick 55
+    indices = replicateM 55 . getStdRandom $ randomR (0,73)
+    -- '()+_,=./+?0123456789:ABCDEFGHIJKLMNOPQRSTUVWXYZabcdefghijklmnopqrstuvwxyz
+    validChars = listArray (0 :: Int, 73)
+                           [ 0x27, 0x28, 0x29, 0x2b, 0x5f, 0x2c, 0x3d, 0x2e
+                           , 0x2f, 0x2b, 0x3f, 0x30, 0x31, 0x32, 0x33, 0x34
+                           , 0x35, 0x36, 0x37, 0x38, 0x39, 0x3a, 0x41, 0x42
+                           , 0x43, 0x44, 0x45, 0x46, 0x47, 0x48, 0x49, 0x4a
+                           , 0x4b, 0x4c, 0x4d, 0x4e, 0x4f, 0x50, 0x51, 0x52
+                           , 0x53, 0x54, 0x55, 0x56, 0x57, 0x58, 0x59, 0x5a
+                           , 0x61, 0x62, 0x63, 0x64, 0x65, 0x66, 0x67, 0x68
+                           , 0x69, 0x6a, 0x6b, 0x6c, 0x6d, 0x6e, 0x6f, 0x70
+                           , 0x71, 0x72, 0x73, 0x74, 0x75, 0x76, 0x77, 0x78
+                           , 0x79, 0x7a
+                           ]
+
+-- | Given a bytestring for the boundary, turns a `MultipartData` into
+-- a 'RequestBody'
+multipartToBody :: forall tag. 
+                MultipartBackend tag
+                => LBS.ByteString
+                -> MultipartData tag
+                -> RequestBody
+multipartToBody boundary mp = RequestBodySource $ files' <> source ["--", boundary, "--"]
+  where
+    -- at time of writing no Semigroup or Monoid instance exists for SourceT and StepT
+    -- in releases of Servant; they are in master though
+    (SourceT l) <> (SourceT r) = SourceT $ \k ->
```

Comment:
I'd use different symbol, it shadows `Prelude`-exported `<>`

Reply:
I wanted migrating over once `Semigroup` and `Monoid` instances are available to be as easy as deleting this section, but I can change the names, np.

Reply:
fixed in 1f4afa4